[LBSE] Add support for 'foreignObject' elements

nikolaszimmermann opened the pull request:
#### 595633178b2f668b50dbb6346265be76c80b11d5
<pre>
[LBSE] Add support for &apos;foreignObject&apos; elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=243498">https://bugs.webkit.org/show_bug.cgi?id=243498</a>

Reviewed by Rob Buis.

Rename RenderSVGForeignObject -&gt; LegacyRenderSVGForeignObject.

Re-introduce RenderSVGForeignObject for LBSE. This marks the finish
of the renderer conversion: all non-resource related renderers have both a legacy
and LBSE Renderer class now: &lt;foreignObject&gt; was the last missing element in LBSE.

Covered by existing svg/foreignObject tests, which all pass now in LBSE (except
three that depends on clipping / masking / filters).

* LayoutTests/platform/mac-monterey-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/background-render-phase-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/body-background-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/clip-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/disallowed-svg-nodes-as-direct-children-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-display-none-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-display-none-with-relative-pos-content-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-display-changes-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-display-none-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-display-none-with-relative-pos-content-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-of-parent-display-none-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-of-parent-display-none-with-relative-pos-content-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/filter-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/mask-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/multiple-foreign-objects-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/no-crash-with-svg-content-in-html-document-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/no-crash-with-svg-content-in-html-document-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/svg-document-as-direct-child-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/svg-document-in-html-document-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/svg-document-in-html-document-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/text-tref-02-b-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/text-tref-02-b-expected.txt: Added.
* LayoutTests/platform/mac/svg/foreignObject/fO-parent-display-changes-expected.txt:
* LayoutTests/svg/foreignObject/fO-parent-display-changes.svg:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AccessibilitySVGElement.cpp:
(WebCore::AccessibilitySVGElement::determineAccessibilityRole):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::didAttachChild):
(WebCore::RenderElement::layerCreationAllowedForSubtree const):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::canContainFixedPositionObjects const):
(WebCore::RenderElement::canContainAbsolutelyPositionedObjects const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateAncestorDependentState):
(WebCore::RenderLayer::paintForegroundForFragments):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::shouldPaintSVGRenderer const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isLegacySVGForeignObject const):
(WebCore::RenderObject::isSVGForeignObjectOrLegacySVGForeignObject const):
* Source/WebCore/rendering/svg/LegacyRenderSVGForeignObject.cpp: Copied from Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp.
(WebCore::LegacyRenderSVGForeignObject::LegacyRenderSVGForeignObject):
(WebCore::LegacyRenderSVGForeignObject::foreignObjectElement const):
(WebCore::LegacyRenderSVGForeignObject::paint):
(WebCore::LegacyRenderSVGForeignObject::localToParentTransform const):
(WebCore::LegacyRenderSVGForeignObject::updateLogicalWidth):
(WebCore::LegacyRenderSVGForeignObject::computeLogicalHeight const):
(WebCore::LegacyRenderSVGForeignObject::layout):
(WebCore::LegacyRenderSVGForeignObject::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/LegacyRenderSVGForeignObject.h: Copied from Source/WebCore/rendering/svg/RenderSVGForeignObject.h.
* Source/WebCore/rendering/svg/LegacyRenderSVGModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::updateFromStyle):
(WebCore::RenderSVGBlock::mapLocalToContainer const):
(WebCore::RenderSVGBlock::pushMappingToContainer const):
* Source/WebCore/rendering/svg/RenderSVGBlock.h:
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::paint):
(WebCore::RenderSVGForeignObject::updateLogicalWidth):
(WebCore::RenderSVGForeignObject::computeLogicalHeight const):
(WebCore::RenderSVGForeignObject::layout):
(WebCore::RenderSVGForeignObject::overflowClipRect const):
(WebCore::RenderSVGForeignObject::updateFromStyle):
(WebCore::RenderSVGForeignObject::applyTransform const):
(WebCore::RenderSVGForeignObject::localToParentTransform const): Deleted.
(WebCore::RenderSVGForeignObject::nodeAtFloatPoint): Deleted.
* Source/WebCore/rendering/svg/RenderSVGForeignObject.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
* Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::createSVGLayerAwareElementSet):
* Source/WebCore/svg/SVGForeignObjectElement.cpp:
(WebCore::SVGForeignObjectElement::createElementRenderer):
(WebCore::SVGForeignObjectElement::rendererIsNeeded):

Canonical link: <a href="https://commits.webkit.org/253816@main">https://commits.webkit.org/253816@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab9dc4c2b83da9d963846bec487f6cb3e94d04b0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87189 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31276 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18017 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/96212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149762 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91165 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29640 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91201 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92805 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23944 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73995 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/79307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27365 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27312 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/14018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2693 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28997 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/36875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28936 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/33295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/svg/SVGForeignObjectElement.cpp:
```diff
@@ -114,8 +119,7 @@ bool SVGForeignObjectElement::rendererIsNeeded(const RenderStyle& style)
     while (ancestor && ancestor->isSVGElement()) {
-        // FIXME: [LBSE] Upstream foreignObject support.
-        if (ancestor->renderer() && ancestor->renderer()->isLegacySVGHiddenContainer())
+        if (ancestor->renderer() && (ancestor->renderer()->isSVGHiddenContainer() || ancestor->renderer()->isLegacySVGHiddenContainer()))
```
My as well introduce a isSVGHiddenContainerOrLegacySVGHiddenContainer method?

nikolaszimmermann replied:
Right, however it would only be this place using it -- I only introduced these helpers so far if there is more than one callsite that needs it.

rwlbuis reviewed (changes requested):
Like the image patch, this is easier to review with two separate patches.

rwlbuis reviewed (approved):
LGTM.

nikolaszimmermann committed:
[LBSE] Add support for 'foreignObject' elements
https://bugs.webkit.org/show_bug.cgi?id=243498

Reviewed by Rob Buis.

Rename RenderSVGForeignObject -> LegacyRenderSVGForeignObject.

Re-introduce RenderSVGForeignObject for LBSE. This marks the finish
of the renderer conversion: all non-resource related renderers have both a legacy
and LBSE Renderer class now: <foreignObject> was the last missing element in LBSE.

Covered by existing svg/foreignObject tests, which all pass now in LBSE (except
three that depends on clipping / masking / filters).

* LayoutTests/platform/mac-monterey-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/background-render-phase-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/body-background-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/clip-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/disallowed-svg-nodes-as-direct-children-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-display-none-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-display-none-with-relative-pos-content-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-display-changes-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-display-none-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-display-none-with-relative-pos-content-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-of-parent-display-none-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-of-parent-display-none-with-relative-pos-content-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/filter-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/mask-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/multiple-foreign-objects-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/no-crash-with-svg-content-in-html-document-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/no-crash-with-svg-content-in-html-document-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/svg-document-as-direct-child-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/svg-document-in-html-document-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/svg-document-in-html-document-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/text-tref-02-b-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/text-tref-02-b-expected.txt: Added.
* LayoutTests/platform/mac/svg/foreignObject/fO-parent-display-changes-expected.txt:
* LayoutTests/svg/foreignObject/fO-parent-display-changes.svg:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AccessibilitySVGElement.cpp:
(WebCore::AccessibilitySVGElement::determineAccessibilityRole):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::didAttachChild):
(WebCore::RenderElement::layerCreationAllowedForSubtree const):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::canContainFixedPositionObjects const):
(WebCore::RenderElement::canContainAbsolutelyPositionedObjects const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateAncestorDependentState):
(WebCore::RenderLayer::paintForegroundForFragments):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::shouldPaintSVGRenderer const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isLegacySVGForeignObject const):
(WebCore::RenderObject::isSVGForeignObjectOrLegacySVGForeignObject const):
* Source/WebCore/rendering/svg/LegacyRenderSVGForeignObject.cpp: Copied from Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp.
(WebCore::LegacyRenderSVGForeignObject::LegacyRenderSVGForeignObject):
(WebCore::LegacyRenderSVGForeignObject::foreignObjectElement const):
(WebCore::LegacyRenderSVGForeignObject::paint):
(WebCore::LegacyRenderSVGForeignObject::localToParentTransform const):
(WebCore::LegacyRenderSVGForeignObject::updateLogicalWidth):
(WebCore::LegacyRenderSVGForeignObject::computeLogicalHeight const):
(WebCore::LegacyRenderSVGForeignObject::layout):
(WebCore::LegacyRenderSVGForeignObject::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/LegacyRenderSVGForeignObject.h: Copied from Source/WebCore/rendering/svg/RenderSVGForeignObject.h.
* Source/WebCore/rendering/svg/LegacyRenderSVGModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::updateFromStyle):
(WebCore::RenderSVGBlock::mapLocalToContainer const):
(WebCore::RenderSVGBlock::pushMappingToContainer const):
* Source/WebCore/rendering/svg/RenderSVGBlock.h:
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::paint):
(WebCore::RenderSVGForeignObject::updateLogicalWidth):
(WebCore::RenderSVGForeignObject::computeLogicalHeight const):
(WebCore::RenderSVGForeignObject::layout):
(WebCore::RenderSVGForeignObject::overflowClipRect const):
(WebCore::RenderSVGForeignObject::updateFromStyle):
(WebCore::RenderSVGForeignObject::applyTransform const):
(WebCore::RenderSVGForeignObject::localToParentTransform const): Deleted.
(WebCore::RenderSVGForeignObject::nodeAtFloatPoint): Deleted.
* Source/WebCore/rendering/svg/RenderSVGForeignObject.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
* Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::createSVGLayerAwareElementSet):
* Source/WebCore/svg/SVGForeignObjectElement.cpp:
(WebCore::SVGForeignObjectElement::createElementRenderer):
(WebCore::SVGForeignObjectElement::rendererIsNeeded):

Canonical link: https://commits.webkit.org/253816@main
